Pass an HTTP Server adapter instead of using express directly

H4-AshleyIngram opened the pull request:
Passing an HTTP Server rather than using express directly decouples the library from express, allowing consumers to use different HTTP implementations. This is useful in a few scenarios:
 
 - If a consumer wants to use a different HTTP library (e.g. Fastify) they won't need both that and express.
  - Consumers can re-use the existing Nest HTTP server, so JSON-RPC endpoints can live on the same port as other Nest endpoints which don't use JSON-RPC (e.g. healthchecks and metrics)
  - Tests can use mock HTTP server implementations like supertest, rather than using a real HTTP server. This is faster, and aids in test isolation (by avoiding port conflicts).

H4-AshleyIngram committed:
Pass an HTTP Server adapter instead of using express directly

Passing an HTTP Server rather than using express directly decouples the library from express, allowing consumers to use different HTTP implementations. This is useful in a few scenarios:
  - If a consumer wants to use a different HTTP library (e.g. Fastify) they won't need both that and express.
  - Consumers can re-use the existing Nest HTTP server, so JSON-RPC endpoints can live on the same port as other Nest endpoints which don't use JSON-RPC (e.g. healthchecks and metrics)
  - Tests can use mock HTTP server implementations like supertest, rather than using a real HTTP server. This is faster, and aids in test isolation (by avoiding port conflicts).

spion commented on README.md:
```diff
@@ -15,15 +15,19 @@ Currently uses HTTP as the transport layer, with plans to add other options as t
 ## Usage example
 
-Initialize similar to a regular microservice, but pass a `JsonRpcService` as the strategy option:
+Initialize similar to a regular microservice used in a hybrid application, but pass a `JsonRpcService` as the strategy option:
```
Can we keep both examples i.e. standalone and hybrid-app?

Something like: " You can also use it in a hybrid application on the same port as a HTTP server by passing the http adapter, like so:"

spion commented on src/server.ts:
```diff
@@ -17,19 +16,16 @@ export class JsonRpcContext {
   /**
-   * Listening port for the HTTP server
-   */
-  port: number;
```
We should probably keep this option as well

spion commented on src/server.ts:
```diff
@@ -66,19 +62,18 @@ export class JsonRpcServer extends Server implements CustomTransportStrategy {
-    app.post(this.options.path, express.json(), async (req, res) => {
-      // let handlers = this.getHandlers();
+    let app = this.options.server ?? express();
+    app.use(bodyParser.json());
```
This will activate a global JSON parser on all routes in the app, which is somewhat invasive.

H4-AshleyIngram committed:
Support standalone and hybrid side-by-side

Support running JSON-RPC services as either a standalone microservice, or as part of a wider Nest application. Add tests to show that both work.
